src/main/kotlin/uk/gov/justice/digital/hmpps/prisonercontactregistry/service/PrisonerContactRegistryService.kt

```diff
@@ -45,6 +55,44 @@ class PrisonerContactRegistryService(private val prisonApiClient: PrisonApiClien
     return contacts
   }
 
+  @Throws(VisitorNotFoundException::class, DateRangeNotFoundException::class)
+  fun getBannedDateRangeForPrisonerContacts(
+    prisonerId: String,
+    visitorIds: List<Long>,
+    fromDate: LocalDate,
+    toDate: LocalDate,
+  ): DateRangeDto {
+    val dateRange = DateRangeDto(fromDate, toDate)
+
+    val contacts = getContactById(prisonerId)
+    val visitors = contacts.filter { visitorIds.contains(it.personId) }
+    if (visitors.size != visitorIds.size) {
+      throw VisitorNotFoundException(message = "Not all visitors provided ($visitorIds) are listed contacts for prisoner $prisonerId")
+    }
+
+    val visitorBanRestrictions = visitors
+      .flatMap { it.restrictions }
+      .filter { it.restrictionType == "BAN" }
+
+    for (restriction in visitorBanRestrictions) {
+      restriction.expiryDate?.let { expiryDate ->
+        if (expiryDate.isAfter(dateRange.toDate)) {
+          throw DateRangeNotFoundException(message = "Found visitor with restriction of 'BAN' with expiry date after our endDate, no date range possible")
+        }
+
+        if (expiryDate.isAfter(dateRange.fromDate)) {
+          dateRange.fromDate = expiryDate
```

Comment:
what if the first expiryDate is later than the last, for example first one is 20/05/2024 and the second is 15/05/2024,... would we loose the more restrictive?

Reply:
We always want to take the most restrictive range right? If I am thinking correctly, it will always pick the date furthest possible...

Example:

initial fromDate: 10/05/24
initial toDate: 25/05/24

2 ban restrictions found
Ban 1 (from: 01/05/24 -> to: 15/05/24)
Ban 2 (from: 02/05/24 -> to: 20/05/24)

The code would output

final fromDate: 20/05/24 (changed to the latest possible ban restriction expiry date found)
from toDate: 25/05/24 (unchanged)

Is that correct?

Reply:
Yeha sorry was in a rush, the code looks good.